### PR TITLE
[FSDP2] Sharded parameter in `FSDPParam`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118584
* #118298
* #118223
* #118198
* #118188
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* #117955
* #117950
* #117881
* __->__ #117877
* #117867
* #117814
* #118525

This PR adds logic to shard the managed parameters on dim-0. This is like `distribute_tensor()` with two differences:
1. `distribute_tensor()` today cannot accept a `DTensor` and reshard it to the parent mesh (https://github.com/pytorch/pytorch/issues/116101).
2. `DTensor` does not pad its local shard on any `Shard` dimensions (https://github.com/pytorch/pytorch/issues/113045).

As such, the `FSDPParam._init_sharded_param()` derives the global `DTensor` metadata itself and pads the local tensor on dim-0. The padding helps make the all-gather copy-in more efficient since the all-gather buffer will require padding.

---

Some details:
- We free the original parameter manually after constructing the sharded parameter. This lowers the peak memory during construction time slightly (since not _all_ parameters in the group must be sharded before the original parameters are freed) and is not strictly necessary.
- We bypass `nn.Module.__setattr__` because the checks are slow and unnecessary. The drawback is that we would ignore a user-defined override of `__setattr__`; however, since we have never encountered this in practice, I am okay with this. Notably, user calls to `setattr` would still use the override; FSDP only uses `setattr` as a mechanism for switching between sharded and unsharded parameters.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 1]
```diff
@@ -20,3 +68,137 @@ def __init__(
...
+            self._global_stride,
+        )
+
+
+# NOTE: Unsafe here refers to not checking whether the storage is already
+# allocated or freed, respectively. We should be safe to use them since we
+# explicitly manage the state transition.
+def unsafe_alloc_storage(tensor: torch.Tensor) -> None:
+    # Skip the already-allocated check and assume that `tensor` is the base
+    # tensor to save CPU overhead
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:187]: are we indicating high cpu times by using
```storage_size: int = tensor.untyped_storage().size()``` ? OR will it trigger cuda async? 

Trying to understand if this is chasing for perf savings or code elegance

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:187]: This is all CPU side overhead. You are correct that it is saving the `tensor.untyped_storage().size()` check(s). E.g., we could check if the size is already the allocated size and not do anything, and we could check if the size is greater than zero but not the desired size and raise an error. We could also check that the tensor is the base tensor / is not a view with nonzero storage offset.

I mainly avoided that because it was not necessary for our use case. I did not want this `unsafe_alloc_storage` to be a general function that others (outside FSDP) would use. If you are using this, you should have a reason for it. For our design, we make sure that we will never call it when the storage size is greater than 0 by guarding with the `ShardedState` state transitions, and we never call it when the tensor is not the base tensor since FSDP manages the tensor that we are passing in.

If people prefer, I can add back those checks because they will not incur too much CPU overhead, so it does not matter much either way.

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 2]
```diff
@@ -39,3 +84,138 @@ def __init__(
...
+            self._global_stride = param.stride()
+            param_data = param
+        self._orig_size = param_data.size()
+        shard_rank = self.mesh_info.shard_mesh_rank
+        shard_world_size = self.mesh_info.shard_mesh_size
+        chunks = _chunk_with_empty(param_data, shard_world_size, dim=0)
+        sharded_param = chunks[shard_rank]
+        self.sharded_size = sharded_param.size()
+        padded_sharded_size = chunks[0].size()  # 0th always padded
+        padded_sharded_param = param_data.new_zeros(padded_sharded_size)
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:153]: I recalled we use the `F.pad` for flat param fsdp, curious why we change to a new + copy_?

File: test/distributed/_composable/fsdp/test_fully_shard_init.py [hunk 3]
```diff
@@ -270,5 +270,81 @@ def test_get_param_module_infos_duplicates(self):
...
+        self.assertEqual(len(orig_params), len(sharded_params))
+        global_mesh = init_device_mesh("cuda", (self.world_size,))
+        for orig_param, sharded_param in zip(orig_params, sharded_params):
+            self.assertIsInstance(sharded_param, DTensor)
+            self.assertEqual(sharded_param.device_mesh, global_mesh)
+            self.assertEqual(sharded_param.size(), orig_param.size())
+            self.assertEqual(sharded_param.stride(), orig_param.stride())
+            self.assertEqual(sharded_param._spec.placements, (Shard(0),))
+            chunks = torch.chunk(orig_param, self.world_size, dim=0)
+            self.assertEqual(sharded_param._local_tensor, chunks[self.rank])
```
Comment (wanchaol) [inline test/distributed/_composable/fsdp/test_fully_shard_init.py:306]: hmmm for uneven sharding, looks like this test should fail?
i.e. `chunks = torch.chunk(orig_param, self.world_size, dim=0)` should be unpadded while the sharded_param._local_tensor should be already padded?

**Review by wanchaol — COMMENTED** [review]
This mostly lgtm, but I think the uneven sharding case aren't tested properly?

File: test/distributed/_composable/fsdp/test_fully_shard_init.py (hunk 3 above)
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_init.py:306]: `sharded_param._local_tensor` is a view into the padded tensor, which FSDP maintains, so `sharded_param._local_tensor` is not padded.

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:153]: `F.pad` is hard to use for ND case (i.e. it is only easy for 1D case). It requires padding from right-to-left, so the args would depend on `ndim`.

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 4]
```diff
@@ -39,3 +84,138 @@ def __init__(
...
+            global_placements: List[Placement] = [Replicate(), Replicate()]
+            global_dp_mesh_dim = _mesh_resources.get_parent_mesh_dim(dp_mesh)
+            global_tp_mesh_dim = _mesh_resources.get_parent_mesh_dim(tp_mesh)
+            assert global_dp_mesh_dim is not None  # mypy
+            assert global_tp_mesh_dim is not None  # mypy
+            # TODO: Hard code FSDP + TP; need to support HSDP + TP
+            global_placements[global_dp_mesh_dim] = Shard(0)
+            global_placements[global_tp_mesh_dim] = self._tp_spec.placements[0]
+            self._global_placements = tuple(global_placements)
+            self._global_size = self._tp_global_size
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:133]: That makes sense. I think we can de-duplicate the `_tp_global_size` and `_tp_global_stride`. You are right that `self._global_size = param.size()` and `self._global_stride = param.stride()`.

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:153]: I think if we use the `Shard(0)` placement in the FSDP DTensor, then we can possibly reuse this function to pad, and it support ND case iiuc https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/placement_types.py#L93

But this can be some follow up to explore

**Review by wanchaol — APPROVED** [review]
lgtm!

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:153]: Thanks for the reference! This function looks useful :)

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 5]
```diff
@@ -39,3 +84,138 @@ def __init__(
...
+        padded_sharded_size = chunks[0].size()  # 0th always padded
+        padded_sharded_param = param_data.new_zeros(padded_sharded_size)
+        if sharded_param.numel() > 0:
+            padded_sharded_param[: sharded_param.size(0)].copy_(sharded_param)
+        self._sharded_param_data = padded_sharded_param.view(-1)
+        self.sharded_param = nn.Parameter(
+            self.to_sharded_dtensor(padded_sharded_param[: sharded_param.size(0)])
+        )
+        self.sharded_param.requires_grad_(param.requires_grad)
+        unsafe_free_storage(param_data)  # free immediately
```
Comment (yifuwang) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:161]: This might be just unnecessary paranoia, but imagine the following scenario::
```python
>>> buf = torch.empty(20, 20)
>>> param_0 = torch.nn.Parameters(buf[:10])
>>> param_1 = torch.nn.Parameter(buf[10:])
>>> param_0.untyped_storage().resize_(0)
# oops we haven't processed param_1 yet
>>> param_1.untyped_storage().size()
0
```

This is probably not a common practice, but it's something the user can do.
